mmgen/core/evaluation/evaluation.py

```diff
@@ -235,66 +236,71 @@ def single_gpu_online_evaluation(model, data_loader, metrics, logger,
     # shared by these metrics. For special metrics like 'PPL', images are
     # generated in a metric-special way and not shared between different
     # metrics.
-    # For probabilistic metrics like 'GaussianKLD', they do not
+    # For reconstruction metrics like 'GaussianKLD', they do not
     # receive images but receive a dict with cooresponding probabilistic
-    # parameter. To make the model return probabilistic
+    # parameter.
 
     special_metrics = []
-    probabilistic_metrics = []
+    recon_metrics = []
     vanilla_metrics = []
     special_metric_name = ['PPL']
-    probabilistic_metric_name = ['GaussianKLD']
+    recon_metric_name = ['GaussianKLD']
     for metric in metrics:
         if metric.name in special_metric_name:
             special_metrics.append(metric)
-        elif metric.name in probabilistic_metric_name:
-            probabilistic_metrics.append(metric)
+        elif metric.name in recon_metric_name:
+            recon_metrics.append(metric)
         else:
             vanilla_metrics.append(metric)
 
-    max_num_images = 0 if len(vanilla_metrics) == 0 else max(
-        metric.num_images for metric in vanilla_metrics)
-    for metric in vanilla_metrics:
+    # define mmcv progress bar
+    max_num_images = 0
+    for metric in vanilla_metrics + recon_metrics:
         metric.prepare()
-        # avoid print empty pbar for metrics do not need reals
-        if metric.num_real_feeded == metric.num_images:
-            continue
-        mmcv.print_log(f'Feed reals to {metric.name} metric.', 'mmgen')
-        pbar = mmcv.ProgressBar(metric.num_real_need)
-        # feed in real images
-        for data in data_loader:
-            # key for unconditional GAN
-            if 'real_img' in data:
-                reals = data['real_img']
-            # key for conditional GAN
-            elif 'img' in data:
-                reals = data['img']
-            else:
-                raise KeyError('Cannot found key for images in data_dict. '
-                               'Only support `real_img` for unconditional '
-                               'datasets and `img` for conditional '
-                               'datasets.')
-
-            if reals.shape[1] not in [1, 3]:
-                raise RuntimeError('real images should have one or three '
-                                   'channels in the first, '
-                                   'not % d' % reals.shape[1])
-            if reals.shape[1] == 1:
-                reals = reals.repeat(1, 3, 1, 1)
-            num_feed = metric.feed(reals, 'reals')
-            if num_feed <= 0:
-                break
+        max_num_images = max(max_num_images,
+                             metric.num_real_need - metric.num_real_feeded)
+    mmcv.print_log(f'Sample {max_num_images} real images for evaluation',
+                   'mmgen')
+    pbar = mmcv.ProgressBar(max_num_images)
+    for data in data_loader:
```

Comment:
There exist many cases where we do not need to construct `data_loader`. In the future, the `data_loader` may be `None`. If you change the logic here, I may suggest that you can add `data_loader = [] if data_loader is None else data_loader` at the start of the function.